integration-test/src/main/java/com/sequenceiq/it/cloudbreak/context/TestContext.java

```diff
@@ -186,6 +186,64 @@ public InstanceAwait getInstanceAwait() {
         return instanceAwait;
     }
 
+    public <E extends Exception, T extends CloudbreakTestDto, U extends MicroserviceClient> T whenException(Class<T> entityClass,
+            Class<? extends MicroserviceClient> clientClass, Action<T, U> action, Class<E> expectedException) {
+        return whenException(entityClass, clientClass, action, expectedException, emptyRunningParameter());
+    }
+
+    public <E extends Exception, T extends CloudbreakTestDto, U extends MicroserviceClient> T whenException(Class<T> entityClass,
+            Class<? extends MicroserviceClient> clientClass, Action<T, U> action, Class<E> expectedException, RunningParameter runningParameter) {
+        return whenException(getEntityFromEntityClass(entityClass, runningParameter), clientClass, action, expectedException, runningParameter);
+    }
+
+    public <E extends Exception, T extends CloudbreakTestDto, U extends MicroserviceClient> T whenException(T entity,
+            Class<? extends MicroserviceClient> clientClass, Action<T, U> action, Class<E> expectedException) {
+        return whenException(entity, clientClass, action, expectedException, emptyRunningParameter());
+    }
+
+    public <E extends Exception, T extends CloudbreakTestDto, U extends MicroserviceClient> T whenException(T entity,
+            Class<? extends MicroserviceClient> clientClass, Action<T, U> action, Class<E> expectedException, RunningParameter runningParameter) {
+        checkShutdown();
+        String key = runningParameter.getKey();
+        if (key == null || key.isEmpty()) {
+            key = action.getClass().getSimpleName();
+        }
+        if (!getExceptionMap().isEmpty() && runningParameter.isSkipOnFail()) {
+            LOGGER.info("Should be skipped because of previous error. when [{}]", key);
+            return entity;
+        }
+
+        CloudbreakUser who = getWho(runningParameter);
+
+        LOGGER.info("when {} action on {} by {}, name: {}", key, entity, who, entity.getName());
+        Log.when(LOGGER, action.getClass().getSimpleName() + " action on " + entity + " by " + who);
+
+        try {
+            return doAction(entity, clientClass, action, who.getAccessKey());
+        } catch (Exception e) {
+            if (!e.getClass().equals(expectedException)) {
```

Comment:
Could you please refactor into a new method and reuse here and in `expect` as well?